File: scripts/cursor-recovery-query.mjs
import fs from 'fs';
import path from 'path';
import os from 'os';

const dbPath = path.join(
  os.homedir(),
  'AppData/Roaming/Cursor/User/globalStorage/state.vscdb'
);

let Database;
try {
  Database = (await import('better-sqlite3')).default;
} catch {
  console.error('better-sqlite3 not installed');
  process.exit(1);
}

const db = new Database(dbPath, { readonly: true });

const patterns = ['%rule%', '%composer%', '%chat%', '%memory%', '%aichat%', '%cursor%'];
const where = patterns.map((p) => `key LIKE '${p}'`).join(' OR ');
const rows = db
  .prepare(`SELECT key, length(value) as len FROM ItemTable WHERE ${where} ORDER BY len DESC LIMIT 60`)
  .all();

console.log('=== KEYS ===');
for (const r of rows) console.log(`${r.len}\t${r.key}`);

const ruleKeys = db
  .prepare(`SELECT key, value FROM ItemTable WHERE key LIKE '%rule%' OR key LIKE '%memory%' LIMIT 20`)
  .all();

console.log('\n=== RULE/MEMORY SAMPLES ===');
for (const r of ruleKeys) {
  const preview = String(r.value).slice(0, 500);
  console.log(`\n--- ${r.key} ---\n${preview}`);
}

db.close();

File: scripts/cursor-recovery-query.py
import os
import sqlite3

p = os.path.join(os.environ["APPDATA"], "Cursor", "User", "globalStorage", "state.vscdb")
c = sqlite3.connect(f"file:{p}?mode=ro", uri=True)
cur = c.cursor()

cur.execute(
    """
    SELECT key, length(value)
    FROM ItemTable
    WHERE key LIKE '%rule%' OR key LIKE '%composer%' OR key LIKE '%chat%'
       OR key LIKE '%memory%' OR key LIKE '%aichat%'
    ORDER BY length(value) DESC
    LIMIT 50
    """
)
print("=== TOP KEYS ===")
for key, ln in cur.fetchall():
    print(f"{ln}\t{key}")

cur.execute(
    """
    SELECT key, value
    FROM ItemTable
    WHERE key LIKE '%userRule%' OR key LIKE '%cursor.rules%'
       OR key LIKE '%memories%' OR key LIKE '%aicontext.personal%'
    LIMIT 20
    """
)
print("\n=== RULE CONTENT ===")
for key, val in cur.fetchall():
    print(f"\n--- {key} ---")
    text = val.decode("utf-8", errors="replace") if isinstance(val, bytes) else str(val)
    print(text[:2000])

c.close()
